rosbag2_storage_mcap/test/rosbag2_storage_mcap/test_mcap_storage.cpp

```diff
@@ -0,0 +1,98 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+#include "rcpputils/filesystem_helper.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#ifdef ROSBAG2_STORAGE_MCAP_HAS_STORAGE_OPTIONS
+#include "rosbag2_storage/storage_options.hpp"
+using StorageOptions = rosbag2_storage::StorageOptions;
+#else
+#include "rosbag2_cpp/storage_options.hpp"
+using StorageOptions = rosbag2_cpp::StorageOptions;
+#endif
+#include "rosbag2_test_common/temporary_directory_fixture.hpp"
+#include "std_msgs/msg/string.hpp"
+
+#include <gmock/gmock.h>
+
+#include <memory>
+#include <string>
+
+using namespace ::testing;  // NOLINT
+using TemporaryDirectoryFixture = rosbag2_test_common::TemporaryDirectoryFixture;
+
+TEST_F(TemporaryDirectoryFixture, can_write_and_read_basic_mcap_file) {
+  auto uri = rcpputils::fs::path(temporary_dir_path_) / "bag";
+  auto expected_bag = uri / "bag_0.mcap";
+  const int64_t timestamp_nanos = 100;  // arbitrary value
+  rcutils_time_point_value_t time_stamp{timestamp_nanos};
+  const std::string topic_name = "test_topic";
+  const std::string message_data = "Test Message 1";
+  const std::string storage_id = "mcap";
+  // NOTE: using cumbersome APIs for Foxy compatibility,
+  // which did not yet provide plain-message API
+  // TODO(emersonknapp) - when Foxy reaches EOL this code can get cut way back
+  rclcpp::Serialization<std_msgs::msg::String> serialization;
+
+  {
+    StorageOptions options;
+    options.uri = uri.string();
+    options.storage_id = storage_id;
+    rosbag2_storage::TopicMetadata topic_metadata;
+    topic_metadata.name = topic_name;
+    topic_metadata.type = "std_msgs/msg/String";
+
+    std_msgs::msg::String msg;
+    msg.data = message_data;
+
+    rosbag2_cpp::Writer writer{std::make_unique<rosbag2_cpp::writers::SequentialWriter>()};
+#ifndef ROSBAG2_STORAGE_MCAP_WRITER_CREATES_DIRECTORY
+    rcpputils::fs::create_directories(uri);
+#endif
+    writer.open(options, rosbag2_cpp::ConverterOptions{});
+    writer.create_topic(topic_metadata);
+
+    auto serialized_msg = std::make_shared<rclcpp::SerializedMessage>();
+    serialization.serialize_message(&msg, serialized_msg.get());
+
+    auto serialized_bag_msg = std::make_shared<rosbag2_storage::SerializedBagMessage>();
+    serialized_bag_msg->serialized_data = std::shared_ptr<rcutils_uint8_array_t>(
+      const_cast<rcutils_uint8_array_t*>(&serialized_msg->get_rcl_serialized_message()),
+      [](rcutils_uint8_array_t* /* data */) {});
```

Comment:
Seems like a misuse of shared_ptr, so probably could at least use a comment explaining why it has to be this way

Reply:
agreed - it's a big misuse of everything. added a comment. when Foxy hits EOL we can really simplify this test.